src/mongo/db/pipeline/expression.cpp

```diff
@@ -2212,6 +2216,38 @@ Value ExpressionFilter::evaluate(const Document& root) const {
     return Value(std::move(output));
 }
 
+Value ExpressionFilter::filterTillNthValue(const Document& root, long long n) const {
```

Comment:
Let's try to de-duplicate some of the code between ExpressionFilter::evaluate and ExpressionFilter::filterTillNthValue.

How about if the second argument here was a boost::optional<long long>, and ExpressionArrayElemAt called this function by passing a long long there (not boost::none), then the contents of ExpressionFilter::evaluate() can entirely be replaced with `filterTillNthValue(root, boost::none)`? Of course, you'd have to change line 2242 a bit, and the name might be a little misleading.

How does that sound?